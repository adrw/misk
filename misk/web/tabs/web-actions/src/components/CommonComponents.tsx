import {
  Button as BlueprintButton,
  Collapse,
  Colors,
  Icon,
  Menu,
  MenuItem,
  Tag,
  TextArea,
  Tooltip as BlueprintTooltip
} from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import { HTTPMethodIntent, WrapTextContainer } from "@misk/core"
import {
  ISimpleFormState,
  onChangeToggleFnCall,
  onClickFnCall,
  simpleSelect
} from "@misk/simpleredux"
import copy from "copy-to-clipboard"
import HTTPMethod from "http-method-enum"
import * as React from "react"
import { connect } from "react-redux"
import styled from "styled-components"
import { IDispatchProps, mapDispatchToProps, mapStateToProps } from "../ducks"

export const FloatLeft = styled.span`
  float: left;
  margin: 5px 10px 0 0;
`

export const FloatRight = styled.span`
  float: right;
  margin: 5px 0 0 10px;
`

export const CodeTag = styled(Tag)`
  font-family: monospace;
`

export const Header = styled.div`
  display: inline-block;
`

export const Column = styled.div`
  flex-grow: 1;
  flex-basis: 0;
  min-width: 320px;
`

export const Button = styled(BlueprintButton)`
  line-height: normal;
  text-transform: inherit;
`

export const Tooltip = styled(BlueprintTooltip)`
  span.bp3-popover-target {
    display: inherit !important;
  }
`

/**
 * Renders HTTP Method tags for each Web Action card
 */
export const MethodTag = (props: { method: HTTPMethod }) => (
  <FloatRight>
    <Tag large={true} intent={HTTPMethodIntent[props.method]}>
      {props.method}
    </Tag>
  </FloatRight>
)

export const MetadataMenu = styled(Menu)`
  li {
    margin-bottom: 0;
  }
`

export const WrapTextArea = styled(TextArea)`
  white-space: pre-line;
`

const definedOrDefault = (item: any, booleanIfUndefined: boolean) =>
  item !== undefined ? item : booleanIfUndefined

/**
 * Single cell component for a piece of metadata
 * @param props
 *  * content: primary content that is displayed (ie. eng, service-owners)
 *  * label: right aligned faded text that is a label for the data (ie. Roles)
 *  * tooltip: text displayed in tooltip when content is hovered over.
 *      Separate definition from content in the case that a truncated length
 *      is displayed but the full length of content is displayed in tooltip
 */
export const Metadata = (props: {
  content: string | JSX.Element
  label?: string
  labelElement?: JSX.Element
  onClick?: (event: any) => void
  tooltip?: string | JSX.Element
}) => {
  if (props.tooltip) {
    return (
      <MenuItem
        {...props}
        text={
          <Tooltip content={props.tooltip} lazy={true}>
            {props.content}
          </Tooltip>
        }
      />
    )
  } else {
    return (
      <MenuItem
        {...props}
        text={<Tooltip lazy={true}>{props.content}</Tooltip>}
      />
    )
  }
}

/**
 * Metadata pre-filled with content to support a click-to-copy-to-clipboard flow
 */
export const MetadataCopyToClipboard = (props: {
  clipboardLabelElement?: boolean
  content?: string | JSX.Element
  data: any
  description?: string
  label?: string
  labelElement?: JSX.Element
  onClick?: (event: any) => void
  tooltip?: string | JSX.Element
}) => {
  const description = props.description && `${props.description} `
  const content: string | JSX.Element =
    props.content || `Click to Copy ${description}to Clipboard`
  const labelElement: JSX.Element =
    props.labelElement ||
    (definedOrDefault(props.clipboardLabelElement, true) && (
      <Tooltip content={"Click to Copy to Clipboard"}>
        <Icon icon={IconNames.CLIPBOARD} />
      </Tooltip>
    ))
  const safeData =
    typeof props.data === "string" ? props.data : JSON.stringify(props.data)
  return (
    <Metadata
      content={content}
      label={props.label}
      labelElement={labelElement}
      onClick={onClickFnCall(copy, safeData)}
      tooltip={props.tooltip}
    />
  )
}

/**
 * MetadataCollapse is a generic container for all Metadata
 * It handles
 *  * Sliding out to reveal child content when clicked
 *  * Automatically showing overflow content below when clicked
 *  * Displaying a list of given content below when clicked
 *  * In most cases all content is "click-to-copy" enabled
 *
 * @param props : includes same props as Metadata with a few additional
 *  * children: any components to display when the Metadata is clicked
 *  * tag: string to use in @misk/SimpleRedux/SimpleForm to register Metadata clicks
 *  * IDispatchProps: include connected dispatch object from parent container
 *      Provides access to @misk/SimpleRedux/SimpleForm input handlers
 */

const UnconnectedMetadataCollapse = (
  props: {
    clipboardLabelElement?: boolean
    children?: any
    content?: string | string[] | JSX.Element | JSX.Element[]
    countLabel?: boolean
    data?: string
    isOpen?: boolean
    label?: string
    labelElement?: JSX.Element
    simpleForm: ISimpleFormState
    tag: string
    text?: string | JSX.Element
    tooltip?: string | JSX.Element
  } & IDispatchProps
) => {
  const content = Array.isArray(props.content) ? props.content : [props.content]
  const collapseIcon = simpleSelect(props.simpleForm, props.tag, "data") ? (
    <Icon icon={IconNames.CARET_DOWN} />
  ) : (
    <Icon icon={IconNames.CARET_RIGHT} />
  )
  return (
    <div>
      <Metadata
        content={
          <span>
            {props.children || content.length > 0 ? (
              collapseIcon
            ) : (
              <Icon color={Colors.LIGHT_GRAY1} icon={IconNames.CARET_RIGHT} />
            )}
            {props.text || props.children || content.join(", ")}
          </span>
        }
        data-testid={"metadata-collapse"}
        label={
          props.children || !props.countLabel
            ? props.label
            : `${props.label} (${content.length})`
        }
        labelElement={props.children ? props.labelElement : null}
        onClick={onChangeToggleFnCall(
          props.simpleFormToggle,
          props.tag,
          props.simpleForm
        )}
      />
      <Collapse isOpen={simpleSelect(props.simpleForm, props.tag, "data")}>
        {props.children
          ? props.children
          : content.map(c => (
              <MetadataCopyToClipboard
                clipboardLabelElement={definedOrDefault(
                  props.clipboardLabelElement,
                  true
                )}
                content={<WrapTextContainer>{c}</WrapTextContainer>}
                data={props.data || c}
                key={c.toString()}
                labelElement={props.labelElement}
              />
            ))}
      </Collapse>
    </div>
  )
}

export const MetadataCollapse = connect(
  mapStateToProps,
  mapDispatchToProps
)(UnconnectedMetadataCollapse)
