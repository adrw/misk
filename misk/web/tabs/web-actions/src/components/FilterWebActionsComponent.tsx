import { ControlGroup, HTMLSelect, InputGroup } from "@blueprintjs/core"
import { onChangeFnCall } from "@misk/simpleredux"
import * as React from "react"
import { IDispatchProps, IState, WebActionInternalLabel } from "../ducks"

export const FilterWebActionsComponent = (
  props: IState & IDispatchProps & { tag: string }
) => {
  const FilterSelectOptions = [...Object.keys(WebActionInternalLabel)]
  const filterTag = `${props.tag}::Filter`
  return (
    <div>
      <ControlGroup fill={true}>
        <HTMLSelect
          large={true}
          onChange={onChangeFnCall(
            props.simpleFormInput,
            `${filterTag}::HTMLSelect`
          )}
          options={FilterSelectOptions}
        />
        <InputGroup
          large={true}
          onChange={onChangeFnCall(
            props.simpleFormInput,
            `${filterTag}::Input`
          )}
          placeholder={"Filter Web Actions"}
        />
        {/* TODO(adrw) Use multiselect to do autocomplete filters */}
      </ControlGroup>
    </div>
  )
}
