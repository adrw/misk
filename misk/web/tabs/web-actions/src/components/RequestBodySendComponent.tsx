import {
  Button,
  Collapse,
  ControlGroup,
  HTMLSelect,
  Label,
  Tag
} from "@blueprintjs/core"
import {
  CodePreContainer,
  HTTPMethodDispatch,
  HTTPMethodIntent
} from "@misk/core"
import {
  onChangeFnCall,
  onClickFnCall,
  simpleSelect,
  simpleType
} from "@misk/simpleredux"
import * as React from "react"
import {
  IDispatchProps,
  IFieldTypeMetadata,
  IState,
  IWebActionInternal
} from "../ducks"
import {
  generateKeyTag,
  parseWebActionConstants,
  safeNumberArray
} from "./utilities"

const selectTypedData = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { requestType, types } = props.action
  const { tag } = props
  if (requestType && types && types[requestType] && types[requestType].fields) {
    const { fields } = types[requestType]
    const fieldsData = fields.map((field: IFieldTypeMetadata) => {
      const fieldKeys = safeNumberArray(
        simpleSelect(
          props.simpleForm,
          generateKeyTag({ field, nestPath: "/", tag }),
          "data",
          simpleType.array
        )
      )
      return fieldKeys.map((key: number) => {
        const keyData = simpleSelect(
          props.simpleForm,
          `${tag}::/${key}::Data`,
          "data"
        )
        console.log(field.name, key, keyData)
        return keyData
      })
    })
    return fieldsData
  } else {
    return []
  }
}

export const RequestBodySendComponent = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { requestType, types } = props.action
  const { url, method, tag } = parseWebActionConstants(props)
  const typedData = selectTypedData(props)
  if (requestType && types && types[requestType] && types[requestType].fields) {
    return (
      <div>
        <ControlGroup>
          <HTMLSelect
            large={true}
            onChange={onChangeFnCall(props.simpleFormInput, `${tag}::Method`)}
            options={props.action.dispatchMechanism.sort()}
            value={method}
          />
          <Button
            large={true}
            onClick={onClickFnCall(
              HTTPMethodDispatch(props)[method],
              `${tag}::Response`,
              url,
              typedData
            )}
            intent={HTTPMethodIntent[method]}
            loading={simpleSelect(
              props.simpleNetwork,
              `${tag}::Response`,
              "loading"
            )}
            text={"Submit"}
          />
        </ControlGroup>
        <Label>
          Request <Tag>{url}</Tag>
        </Label>
        <Collapse isOpen={typedData.length > 0}>
          <CodePreContainer>
            {JSON.stringify(typedData, null, 2)}
          </CodePreContainer>
        </Collapse>
      </div>
    )
  } else {
    return (
      <div>
        <ControlGroup>
          <HTMLSelect
            large={true}
            onChange={onChangeFnCall(props.simpleFormInput, `${tag}::Method`)}
            options={props.action.dispatchMechanism.sort()}
            value={method}
          />
          <Button
            large={true}
            onClick={onClickFnCall(
              HTTPMethodDispatch(props)[method],
              `${tag}::Response`,
              url,
              simpleSelect(props.simpleForm, `${tag}::Body`, "data")
            )}
            intent={HTTPMethodIntent[method]}
            loading={simpleSelect(
              props.simpleNetwork,
              `${tag}::Response`,
              "loading"
            )}
            text={"Submit"}
          />
        </ControlGroup>
        <Label>
          Request <Tag>{url}</Tag>
        </Label>
        <Collapse
          isOpen={simpleSelect(props.simpleForm, `${tag}::Body`, "data")}
        >
          <CodePreContainer>
            {JSON.stringify(
              simpleSelect(props.simpleForm, `${tag}::Body`, "data"),
              null,
              2
            )}
          </CodePreContainer>
        </Collapse>
      </div>
    )
  }
}
