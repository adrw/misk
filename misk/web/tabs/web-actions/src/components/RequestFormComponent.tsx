import {
  Button,
  Card,
  ControlGroup,
  FormGroup,
  InputGroup,
  Intent,
  TextArea,
  Tooltip
} from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import {
  onChangeFnCall,
  onChangeToggleFnCall,
  simpleSelect,
  simpleType
} from "@misk/simpleredux"
import { Map } from "immutable"
import * as React from "react"
import styled from "styled-components"
import {
  BaseFieldTypes,
  findIndexAction,
  IDispatchProps,
  IState,
  ITypesFieldMetadata,
  IWebActionInternal,
  padId,
  TypescriptBaseTypes
} from "../ducks"

const RequestFieldGroup = styled(Card)`
  box-shadow: none !important;
  border-left: 2px black;
  margin-bottom: 0px;
  padding: 0px 0px 0px 10px !important;
`

const RequestFormGroup = styled(FormGroup)`
  margin: 0 !important;
`

const AddButton = (
  props: {
    idParent: string
    action: IWebActionInternal
  } & IState &
    IDispatchProps
) => (
  <Tooltip content={"Add Field"}>
    <Button
      icon={IconNames.PLUS}
      onClick={onChangeFnCall(
        props.webActionsAdd,
        props.idParent,
        props.action,
        props.webActionsRaw
      )}
    />
  </Tooltip>
)

const RemoveButton = (
  props: {
    id: string
    action: IWebActionInternal
  } & IState &
    IDispatchProps
) => (
  <Tooltip content={"Remove Field"}>
    <Button
      icon={IconNames.CROSS}
      onClick={onChangeFnCall(
        props.webActionsRemove,
        props.id,
        props.action,
        props.webActionsRaw
      )}
    />
  </Tooltip>
)

const repeatableFieldButtons = (
  props: {
    action: IWebActionInternal
    id: string
    typesMetadata: Map<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => {
  const { action, id, typesMetadata } = props
  const metadata = typesMetadata.get(id)
  if (
    metadata &&
    metadata.id !== "0" &&
    typesMetadata.get(metadata.idParent).repeated
  ) {
    const { idParent, name, serverType } = metadata
    const { idChildren } = typesMetadata.get(idParent)
    const removeButton =
      idChildren.size > 1 ? (
        <RemoveButton {...props} action={action} id={id} />
      ) : (
        <span />
      )
    return [
      <Tooltip content={`Repeated ${serverType}`}>
        <Button icon={IconNames.REPEAT}>{name}</Button>
      </Tooltip>,
      <AddButton {...props} action={action} idParent={idParent} />,
      removeButton
    ]
  } else if (metadata && !metadata.repeated) {
    const { name, serverType } = metadata
    return [
      <Tooltip content={serverType}>
        <Button>{name || "Body"}</Button>
      </Tooltip>
    ]
  } else {
    return [<span />]
  }
}

const RequestFormFieldBuilder = (
  props: {
    action: IWebActionInternal
    id: string
    tag: string
    typesMetadata: Map<string, ITypesFieldMetadata>
  } & IState &
    IDispatchProps
) => {
  const { id, tag, typesMetadata } = props
  const metadata = typesMetadata.get(id)
  const {
    idChildren,
    name,
    serverType,
    typescriptType
  } = metadata as ITypesFieldMetadata
  if (typescriptType === null) {
    if (
      idChildren.first() &&
      typesMetadata.get(idChildren.first()).typescriptType &&
      BaseFieldTypes.hasOwnProperty(serverType)
    ) {
      return (
        <div>
          {idChildren.map((child: string) => (
            <RequestFormFieldBuilder {...props} id={child} />
          ))}
        </div>
      )
    } else {
      const fieldGroup = (child: string) => {
        const { serverType: childServerType } = typesMetadata.get(child)
        if (BaseFieldTypes.hasOwnProperty(childServerType)) {
          return (
            <div>
              <RequestFormFieldBuilder {...props} id={child} />
            </div>
          )
        } else {
          return (
            <div>
              <ControlGroup>
                {...repeatableFieldButtons({ ...props, id: child })}
              </ControlGroup>
              <RequestFormFieldBuilder {...props} id={child} />
            </div>
          )
        }
      }
      return (
        <RequestFieldGroup>
          {idChildren.map((child: string) => fieldGroup(child))}
        </RequestFieldGroup>
      )
    }
  } else if (typescriptType === TypescriptBaseTypes.boolean) {
    return (
      <ControlGroup>
        {...repeatableFieldButtons({ ...props, id })}
        <Button
          intent={
            simpleSelect(
              props.simpleForm,
              `${tag}::${padId(id)}`,
              "data",
              simpleType.boolean
            )
              ? Intent.PRIMARY
              : Intent.WARNING
          }
          onClick={onChangeToggleFnCall(
            props.simpleFormToggle,
            `${tag}::${padId(id)}`,
            props.simpleForm
          )}
        >
          {simpleSelect(
            props.simpleForm,
            `${tag}::${padId(id)}`,
            "data",
            simpleType.boolean
          ).toString()}
        </Button>
      </ControlGroup>
    )
  } else if (typescriptType === TypescriptBaseTypes.number) {
    return (
      <ControlGroup>
        {...repeatableFieldButtons({ ...props, id })}
        <InputGroup
          onChange={onChangeFnCall(
            props.simpleFormInput,
            `${tag}::${padId(id)}`
          )}
          placeholder={serverType}
          value={simpleSelect(props.simpleForm, `${tag}::${padId(id)}`, "data")}
        />
      </ControlGroup>
    )
  } else if (typescriptType === TypescriptBaseTypes.string) {
    return (
      <ControlGroup>
        {...repeatableFieldButtons({ ...props, id })}
        <Tooltip content={"Toggle large text input"}>
          <Button
            active={simpleSelect(
              props.simpleForm,
              `${tag}::LongText:${padId(id)}`,
              "data",
              simpleType.boolean
            )}
            icon={IconNames.MORE}
            onClick={onChangeToggleFnCall(
              props.simpleFormToggle,
              `${tag}::LongText:${padId(id)}`,
              props.simpleForm
            )}
          />
        </Tooltip>
        {simpleSelect(
          props.simpleForm,
          `${tag}::LongText:${padId(id)}`,
          "data",
          simpleType.boolean
        ) ? (
          <TextArea
            fill={true}
            onChange={onChangeFnCall(
              props.simpleFormInput,
              `${tag}::${padId(id)}`
            )}
            placeholder={`${serverType}\nDrag bottom right corner of text area input to expand.`}
            value={simpleSelect(
              props.simpleForm,
              `${tag}::${padId(id)}`,
              "data"
            )}
          />
        ) : (
          <InputGroup
            onChange={onChangeFnCall(
              props.simpleFormInput,
              `${tag}::${padId(id)}`
            )}
            placeholder={serverType}
            value={simpleSelect(
              props.simpleForm,
              `${tag}::${padId(id)}`,
              "data"
            )}
          />
        )}
      </ControlGroup>
    )
  } else {
    return (
      <div>
        {...repeatableFieldButtons({ ...props, id })}
        <RequestFormGroup label={`${name}:${JSON.stringify(serverType)}`}>
          <TextArea
            fill={true}
            onChange={onChangeFnCall(
              props.simpleFormInput,
              `${tag}::${padId(id)}`
            )}
            placeholder={
              "Unparseable type. (JSON or Text).\nDrag bottom right corner of text area input to expand."
            }
          />
        </RequestFormGroup>
      </div>
    )
  }
}

export const RequestFormComponent = (
  props: {
    action: IWebActionInternal
    tag: string
  } & IState &
    IDispatchProps
) => {
  const { action } = props
  const { typesMetadata } = props.webActionsRaw.get("metadata")[
    findIndexAction(action, props.webActionsRaw)
  ]
  return (
    <RequestFormFieldBuilder
      {...props}
      id={"0"}
      typesMetadata={typesMetadata}
    />
  )
}
