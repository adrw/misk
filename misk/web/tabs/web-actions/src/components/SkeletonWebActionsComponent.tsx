import { Card, Classes, H3 } from "@blueprintjs/core"
import { FlexContainer } from "@misk/core"
import { HTTPMethod } from "http-method-enum"
import * as React from "react"
import {
  CodeTag,
  Column,
  FloatLeft,
  Header,
  Metadata,
  MetadataMenu,
  MethodTag
} from "."

/**
 * Empty Web Action Card UI for use with BlueprintJS Skeleton class in loading UIs
 * https://blueprintjs.com/docs/#core/components/skeleton
 */
export const SkeletonText = () => (
  <span className={Classes.SKELETON}>{"Lorem ipsum"}</span>
)

export const SkeletonWebActionsComponent = () => (
  <Card>
    <Header>
      {[HTTPMethod.GET].map((m, index) => (
        <MethodTag key={index} method={m} />
      ))}
      <FloatLeft>
        <H3 className={Classes.SKELETON}>{"AnotherSimpleWebAction"}</H3>
      </FloatLeft>
      <FloatLeft>
        <CodeTag large={true}>{<SkeletonText />}</CodeTag>
      </FloatLeft>
    </Header>
    <FlexContainer>
      <Column>
        <MetadataMenu>
          <Metadata label={"Function"} content={<SkeletonText />} />
          <Metadata label={"Services"} content={<SkeletonText />} />
          <Metadata label={"Roles"} content={<SkeletonText />} />
          <Metadata label={"Access"} content={<SkeletonText />} />
        </MetadataMenu>
      </Column>
      <Column>
        <MetadataMenu>
          <Metadata label={"Content Types"} content={<SkeletonText />} />
          <Metadata
            label={"Application Interceptors"}
            content={<SkeletonText />}
          />
          <Metadata label={"Network Interceptors"} content={<SkeletonText />} />
          <Metadata label={"Send a Request"} content={<SkeletonText />} />
        </MetadataMenu>
      </Column>
    </FlexContainer>
  </Card>
)
