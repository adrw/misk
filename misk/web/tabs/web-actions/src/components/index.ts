export * from "./CommonComponents"
export * from "./FilterWebActionsComponent"
export * from "./RequestFormComponent"
export * from "./SendRequestCollapseComponent"
export * from "./WebActionsComponent"
