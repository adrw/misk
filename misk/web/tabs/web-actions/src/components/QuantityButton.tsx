import { IconName } from "@blueprintjs/icons"
import { onChangeFnCall } from "@misk/simpleredux"
import * as React from "react"
import { IWebActionInternal } from "../ducks"
import { Button, Tooltip } from "./CommonComponents"

export const QuantityButton = (props: {
  id: string
  action: IWebActionInternal
  changeFn: any
  content: string
  icon: IconName
  oldState: any
}) => (
  <Tooltip content={props.content} lazy={true}>
    <Button
      icon={props.icon}
      onClick={onChangeFnCall(
        props.changeFn,
        props.id,
        props.action,
        props.oldState
      )}
    />
  </Tooltip>
)
