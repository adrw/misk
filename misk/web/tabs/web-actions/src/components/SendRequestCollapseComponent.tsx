import {
  Collapse,
  Icon,
  InputGroup,
  Intent,
  Label,
  Tag
} from "@blueprintjs/core"
import { IconNames } from "@blueprintjs/icons"
import { CodePreContainer, HTTPStatusCodeIntent } from "@misk/core"
import {
  onChangeFnCall,
  onChangeToggleFnCall,
  simpleSelect
} from "@misk/simpleredux"
import * as React from "react"
import { RequestFormComponent } from "../components"
import { IDispatchProps, IState, IWebActionInternal } from "../ducks"
import { RequestBodySendComponent } from "./RequestBodySendComponent"
import { parseWebActionConstants } from "./utilities"

/**
 * Collapse wrapped Send a Request form for each Web Action card
 */
export const SendRequestCollapseComponent = (
  props: { action: IWebActionInternal; tag: string } & IState & IDispatchProps
) => {
  const { isOpen, methodHasBody, tag } = parseWebActionConstants(props)
  return (
    <Collapse isOpen={isOpen}>
      <InputGroup
        defaultValue={props.action.pathPattern}
        onChange={onChangeFnCall(props.simpleFormInput, `${tag}::URL`)}
        placeholder={
          "Request URL: absolute ( http://your.url.com/to/send/a/request/to/ ) or internal service endpoint ( /service/web/action )"
        }
        type={"url"}
      />
      <Collapse isOpen={methodHasBody}>
        <RequestFormComponent {...props} tag={tag} />
        <br />
      </Collapse>
      <RequestBodySendComponent {...props} tag={tag} />
      <Collapse
        isOpen={simpleSelect(props.simpleNetwork, `${tag}::Response`, "status")}
      >
        <Label>
          Response{" "}
          <Tag
            intent={HTTPStatusCodeIntent(
              simpleSelect(props.simpleNetwork, `${tag}::Response`, "status")[0]
            )}
          >
            {(
              simpleSelect(props.simpleNetwork, `${tag}::Response`, "status") ||
              []
            ).join(" ")}
          </Tag>{" "}
          <Tag
            intent={Intent.NONE}
            onClick={onChangeToggleFnCall(
              props.simpleFormToggle,
              `${tag}::ButtonRawResponse`,
              props.simpleForm
            )}
          >
            <span>
              Raw Response{" "}
              {simpleSelect(
                props.simpleForm,
                `${tag}::ButtonRawResponse`,
                "data"
              ) ? (
                <Icon icon={IconNames.CARET_DOWN} />
              ) : (
                <Icon icon={IconNames.CARET_RIGHT} />
              )}
            </span>
          </Tag>
        </Label>
      </Collapse>
      <Collapse
        isOpen={simpleSelect(props.simpleNetwork, `${tag}::Response`, "data")}
      >
        <CodePreContainer>
          {JSON.stringify(
            simpleSelect(props.simpleNetwork, `${tag}::Response`, "data"),
            null,
            2
          )}
        </CodePreContainer>
      </Collapse>
      <Collapse
        isOpen={simpleSelect(
          props.simpleForm,
          `${tag}::ButtonRawResponse`,
          "data"
        )}
      >
        <Label>Raw Network Redux State</Label>
        <CodePreContainer>
          {JSON.stringify(
            simpleSelect(props.simpleNetwork, `${tag}::Response`),
            null,
            2
          )}
        </CodePreContainer>
      </Collapse>
    </Collapse>
  )
}
