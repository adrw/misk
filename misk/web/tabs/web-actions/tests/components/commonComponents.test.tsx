import React from "react"
import { cleanup, render } from "react-testing-library"
import { Metadata, MetadataCollapse, MetadataMenu } from "../../src/components"
import { renderWithRedux } from "../upstreamableTestUtilities"

describe("Common Components", () => {
  afterEach(cleanup)
  it("MetadataMenu", () => {
    const { asFragment } = render(<MetadataMenu />)
    expect(asFragment()).toMatchSnapshot()
  })
  it("Metadata", () => {
    const { asFragment } = render(<Metadata content={"Test Content"} />)
    expect(asFragment()).toMatchSnapshot()
  })
  it("MetadataCollapse can render with redux", () => {
    const { asFragment } = renderWithRedux(
      <MetadataCollapse content={"Test Content"} tag={"TestTag"}>
        <span>{"Collapse Span"}</span>
      </MetadataCollapse>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
