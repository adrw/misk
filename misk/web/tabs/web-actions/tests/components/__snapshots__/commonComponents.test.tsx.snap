// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common Components Metadata 1`] = `
<DocumentFragment>
  <li
    class=""
  >
    <a
      class="bp3-menu-item bp3-popover-dismiss"
      content="Test Content"
    >
      <div
        class="bp3-text-overflow-ellipsis bp3-fill"
      >
        <span
          class="bp3-popover-wrapper sc-bZQynM hRiALA"
        >
          <span
            class="bp3-popover-target"
          >
            <span
              class=""
              tabindex="0"
            >
              Test Content
            </span>
          </span>
        </span>
      </div>
    </a>
  </li>
</DocumentFragment>
`;

exports[`Common Components MetadataCollapse can render with redux 1`] = `
<DocumentFragment>
  <div>
    <li
      class=""
    >
      <a
        class="bp3-menu-item bp3-popover-dismiss"
        content="[object Object]"
        data-testid="metadata-collapse"
      >
        <div
          class="bp3-text-overflow-ellipsis bp3-fill"
        >
          <span
            class="bp3-popover-wrapper sc-bZQynM hRiALA"
          >
            <span
              class="bp3-popover-target"
            >
              <span
                class=""
                tabindex="0"
              >
                <span
                  class="bp3-icon bp3-icon-caret-right"
                  icon="caret-right"
                >
                  <svg
                    data-icon="caret-right"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      caret-right
                    </desc>
                    <path
                      d="M11 8c0-.15-.07-.28-.17-.37l-4-3.5A.495.495 0 0 0 6 4.5v7a.495.495 0 0 0 .83.37l4-3.5c.1-.09.17-.22.17-.37z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span>
                  Collapse Span
                </span>
              </span>
            </span>
          </span>
        </div>
      </a>
    </li>
    <div
      class="bp3-collapse"
    >
      <div
        aria-hidden="false"
        class="bp3-collapse-body"
        style="transform: translateY(-0px);"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Common Components MetadataMenu 1`] = `
<DocumentFragment>
  <ul
    class="bp3-menu sc-gzVnrw ipVOmx"
  />
</DocumentFragment>
`;
